### DISCUSSION TITLE
And now for the Wheel...

### DISCUSSION THREAD

**Discussion by ej113** [tloimu/adapt-ffb-joy#54] [Ideas]
So I've been playing Assetto Corsa (AC) with the gameport Sidewinder Wheel and full force feedback function!

I set myself the goal of getting it working in AC to start with for a number of reasons:

- AC has very good force feedback despite only using a very limited repertoire of force feedback effects and effect updates. It uses damper when stationary which fades out at ~10km/hr and then only constant force above this.
- The streaming of constant force from the physics engine at up to 333Hz is a real test of the adapter & wheel capabilities. 
- It has a similar force feedback implementation to a number of other racing sims/games. There's a good chance the same minimal implementation will also work in Assetto Corsa Competizione, iRacing, some of the F1 series and KartKraft based on evidence from  [this list](https://github.com/Ultrawipf/OpenFFBoard/issues/16#issuecomment-810244245) and [this thread](https://www.assettocorsa.net/forum/index.php?threads/ac-ffb-implementation-improvement-suggestions.1635/), but only if they are using exactly the same set of DirectInput commands.
- Early games with force feedback (like Need For Speed III) tend to use layered "canned" effects with a wider variety of DirectInput effects and parameters, rather than just streaming forces. These will take more effort to get working.
- Finally my aging PC just about meets the minimum spec!

I only had to implement the commands for:

- constant force with a variable signed magnitude, adjustable on-the-fly
- damper with variable positive and negative coefficients, adjustable on-the-fly

But I also got working, because it was straightforward:	

- Spring and inertia with variable positive and negative coefficients - these effects have the same formulation as damper
- All device control commands
- Device gain

The framework for SysEx handling, effect modification and effect control commands for the wheel were already written by Skoo. Also effect duration. Not to mention that It wouldn't have been possible to get this far without his notes in the [Wiki](https://github.com/tloimu/adapt-ffb-joy/blob/wiki/SidewinderFFBWheel.md) and the [DescentBB thread](https://www.descentbb.net/viewtopic.php?t=19061&sid=929f9c7cc0fe125454e1e49561de166c&start=120), since I don't have a Win98 PC with gameport.

The one puzzle I had to solve was in understanding the bit flag 0x00/0x40 which Skoo referred to as a "Default" indicator. It turns out this is not optional and is actually related to the checksum. If (256 < sum of all bytes <=384) then 0x40 else 0x00. This took me quite a while to figure out and sent me off on the wrong track a few times, since some effect modifications  would work and some formulated the same way would not. I do not know why Microsoft made it this way. My best guess is that it was a fudge to avoid interfering with other MIDI devices, but it may just be obfuscation…

### Device Control and Device Gain
These are the only other additions to record for now since everything else I've used is consistent with the wiki.
0xf3 0x1d Reset (stop all effects; free all effects; reset device gain to 0x7F; enable actuators; enable auto spring centre)
0xf3 0x2e Enable actuators
0xf3 0x3f Disable actuators (but time stepping continues for effects)
0xf3 0x48 Pause (time stepping is paused)
0xf3 0x59 Continue (resume paused effects)
0xf3 0x6a Stop All (including stop auto centre)

The last nibble is checksum from xor'ing the first 3 nibbles. Note 4 and 5 are swapped relative to the joystick.

Device Gain is a 6 byte 0xF1 modify message form with effect id  = 0x00, address 0x-0 and value 0x00 to 0x7f in the LSB. e.g. 0xf1, 0x10, 0x40, 0x00, 0x7f, 0x00 for max gain

### USB Descriptors, Axes and Buttons
I have not fiddled much with the descriptors to change the axes/buttons etc. I have fixed the report outputs so the FORCE toggle does not show up as a button and there are no spurious hat switch outputs. There is a bug not yet fixed that means the descriptors have to be the same number of bytes for the joystick and wheel so not much scope to customise this at present. This may result in compatibility issues with some games that aren't very flexible with controller assignments.

### At last some firmware:
[I've put four firmware files here](https://github.com/ej113/adapt-ffb-joy/tree/micro_pinout_wheeltest2_LOG/downloads), options for both Teensy and Micro, with and without virtual serial port logging. These are alpha1 version since I've not done extensive testing. I'm also thinking to keep this out of the main repo until there is a good hardware solution for the backpowering issue #49.

See [here for Teensy wiring](https://github.com/tloimu/adapt-ffb-joy/blob/wiki/HowToBuild.md) and note for Micro/Pro Micro PB6 is swapped for PB0.

If you pick a LOG version you will always need to have a terminal open to the COM port as per #50. You will be able to record the force feedback messages being received and sent over USB and MIDI. Performance doesn't seem to be noticeably worse.

In AC I needed to tick the "Half FFB Update Rate" box because the adapter doesn't seem to handle the default update rate of 333Hz. You get a "99% CPU Occupancy" warning in the game when it is not keeping up. 

There are guides around on how to configure AC FFB for old wheels. I set steering gamma to 1.5 for stability with the limited rotation.

### And a Warning:
**I believe there is potential to damage the Atmega or even the wheel if the adapter is connected to a powered wheel without usb** power. See #49. So far mine have survived despite several prolonged periods in this condition. **USE AT YOUR OWN RISK**

Observe the following connection sequence:
Pedals can be connected at any point

1. Connect usb cable to adapter and check board LED is flashing ~3 times a second
2. Connect 12V power to wheel and wait for it to complete the initialisation - wheel stops turning. If you connect to adapter too early the wheel will be off centre.
3. Connect gameport plug to adapter. You should then get solid LED until the USB "ping" and the LED turns off.

When stopping, always remove 12V power before USB. I recommend not leaving 12V power attached in case your PC stops providing USB power.

### Next Steps
In order to get to a more complete PID implementation like for the joystick it would be necessary to understand how to translate the formula for spring/damper/inertia effects from coefficients, saturations, deadband and offset, into the different representation in the wheel MIDI effects.

There are a few other unknown parameters to discover. Does the wheel have e.g. an explicit effect gain or triggerbutton?

There is a mystery modify message sent by the drivers after each effect is defined by SysEx e.g. 
f1 01 0e 02 7f 7f for periodics. This isn't essential to get effects to play. Perhaps it exposes some additional parameters?

I'm intending to record any further developments on wheel FFB in this thread.

**Comment by You69Man** [comment]
That' great news! I dug out my steering wheel, put it back together with my teensy breadboard that I built a few years ago, and tried two old forcefeedback test programs, WheelCheck.exe and ForceTest.exe. And voila - the device is recognized as "LUFA Wheel wFBB".
And there is actually some force feedback effects working, at least much more than with the version that i had back then!

I'd like to do some more experiments, maybe try some real game. Can you recommend programs or some free game for testing? I don't want to buy a game as long as I don't know if I will be happy with it. 25 years ago my favority game was Sega Rally in Windows 95, but I'm afraid there is no chance for a revival of this great game, not even on Steam or on an emulator...

**Reply by ej113** [reply]
Glad it's working for you. As I said it is really a very limited implementation "proof of concept" targeted at one game and with the possibility that others have near identical force feedback implementations. Keys for Assetto Corsa are pretty cheap these days, but it is a sim more than a game. (I have no particular preference but it was the way to get the most complex force feedback with the least effort on the firmware). It won't behave correctly with older or arcade style games that make more use of canned effects rather than try to simulate the physics.

Re: Sega Rally, it looks like there are people who have got this working with modern FF wheels...

**Comment by ej113** [comment]
For three of the conditional effects, Spring, Damper and Inertia (but not Friction), the response curve in MIDI is set using four pairs of coordinates (xi, yi), with one byte per value. The default is 0x007d 0x3e3f 0x3e3f 0x7d00. This corresponds to a linear response to position, velocity or acceleration respectively:
![image](https://github.com/tloimu/adapt-ffb-joy/assets/132016173/89fd410b-785d-46f0-8ecf-9ef095fafe44)

A position of 0x3f = 63 appears to be about centered
A force level of 0x3f = 63 seems to be about neutral i.e. zero force applied by the motor.
I'm not sure whether it is more correct to use the full range of values to 0x00..0x7f rather than ..0x7d. Perhaps the original drivers were rounding or it was intentionally reduced. 0x7f works as you would expect but it's difficult to tell whether it results in any more force than 0x7d.

Positive and negative coefficients, saturation, offset and deadband can be altered with the coordinates.
e.g. 0x1450 0x2f3f 0x3f0x3f 0x6b1b
![image](https://github.com/tloimu/adapt-ffb-joy/assets/132016173/ac33bac6-cdd3-4cb0-b50a-7d635c1c1197)

Extrapolation is flat so that e.g. position inputs below x1 result in a constant force of y1. So the curve shown above saturates.

By manipulating the coordinates a variety of effects can be generated that are not supported in the USB standard. For example a wall effect where the force goes from zero to maximum over a very short travel.

This doesn't reflect any recent work, but something I figured out a while ago and hadn't got round to appending here. If I recall, most of the remaining parameters are well explained in the Wiki and worked when I tested them. I don't think there are any significant further unknowns to complete an implementation to the same level as the joystick.

**Comment by kalimahhh** [comment]
Hi, I don't know basically nothing about coding but I would like to start playing with my wheel that I own since I bought it 98 or something... I Would like to help in some way giving some feedback on the firmware. 
Should I buy a teensy 2.0 to start and follow the diagram? Thanks for the effort!
I don't want to to throw my sidewinder wheel in the garbage!...

**Reply by ej113** [reply]
Feedback is welcome although note the scope of the wheel implementation is limited so far!

I would suggest USB-C Pro Micro or perhaps Adafruit Itsy-Bitsy 32u4 - 5V 16MHz rather than the Teensy 2.0 unless you have one already, since clones of the Teensy won't have the original proprietary bootloader. But you will need to pay attention to the pinout. There is a wiring diagram for the Pro Micro in this repo https://github.com/juchong/adapt-ffb-joy

**Comment by BrainRadioactivity** [comment]
Just wanted to comment that i tried Asseto Corsa with your new patch and it worked perfectly. It's so wonderful to see my old wheel alive again on a recent OS. I had the exact same experience you described, with needing to set  "Half FFB Update Rate"  to avoid the  "99% CPU Occupancy" warning, but otherwise a very pleasant experience in the game.

I'm gonna keep an eye on this topic, just know that your work is very appreciated ! If you need anything (beta testing, ...), just ask.

**Comment by You69Man** [comment]
I also want to report, that I'm using Teensy 2.0 with the alpha1 binary successfully with Asseto Corsa now, after I have recognized that I need to tick the "Half FFB Update Rate" option! Otherwise my PC had a total hang up.
Is there already any experience with WRC 9?
Thanks again to all who still support our old wheel! Mine is 26 years old meanwhile, and I'm very happy that I kept it so long!

PS: @ej113 Can you tell me which branch/commit in your repo was the base for the alpha1 binary (I guess it's been the micro_pinout_wheeltest2_LOG, right?)? 
And is there any chance for a newer version?

**Comment by Harbor-Gaming** [comment]
Yo, did you need to do any modifications to the adapters  hardware for the microsoft sidewinder force feedback wheel to to work with ffb? im interested in building this but i really want to be sure it will work before i buy the components...

**Comment by You69Man** [comment]
No modification to the recommended adapters hardware and also no modification to the wheel hardware.

Here some pics of my self made adapter:

![PXL_20250509_062146384](https://github.com/user-attachments/assets/16e60ad2-f153-4c13-a579-2ec5cc7b6136)

![PXL_20250509_062149878](https://github.com/user-attachments/assets/d1b5414d-3881-4962-a61e-236e6e4454f9)

![PXL_20250509_062134213](https://github.com/user-attachments/assets/585758a9-a884-4045-b77d-bbf8dd244f8e)

**Reply by Harbor-Gaming** [reply]
Nice, i might build one for myself then. FINALLY i will have that microsoft sidewinder ffb wheel working WITH ffb. Thank you for your attention!

**Reply by Harbor-Gaming** [reply]
Hey man, i finally got the last 2 pieces i needed for the adapter, and my brother already built it with me (he's better with soldering and electronics than me). I flashed it with the "adaptffbjoy-0.5.0 beta1.hex" file, but as soon as i connect the teensy 2.0 to the pc (not in the programming mode) the pc doesnt detect it, and the teensy's light keeps flashing quickly... idk man i already tried so many different .hex release files, all of them keep the teensy flashing, and the pc only detects it when in programming mode. BTW: the wheel detects the teensy eventhough it keeps crashing (i believe that's what the yellow light blinking means) and i can turn the "force" button on or off, if i turn it on the wheel gets auto centered. though that happens only when its conected to the teensy and even if the teensy's off. COULD ANYONE HELP ME PLEASE?

**Reply by You69Man** [reply]
> Hey man, i finally got the last 2 pieces i needed for the adapter, and my brother already built it with me (he's better with soldering and electronics than me). I flashed it with the "adaptffbjoy-0.5.0 beta1.hex" file, but as soon as i connect the teensy 2.0 to the pc (not in the programming mode) the pc doesnt detect it, and the teensy's light keeps flashing quickly... idk man i already tried so many different .hex release files, all of them keep the teensy flashing, and the pc only detects it when in programming mode. BTW: the wheel detects the teensy eventhough it keeps crashing (i believe that's what the yellow light blinking means) and i can turn the "force" button on or off, if i turn it on the wheel gets auto centered. though that happens only when its conected to the teensy and even if the teensy's off. COULD ANYONE HELP ME PLEASE?

The teensy's LED flashing quickly isn't a bad sign. The same does my device after I connect it to USB, and before I connect the wheel into the 15 pin D-Sub connector. I think it only means something like "waiting for the wheel to be connected...".
As soon I connect (the already powered on) wheel to the D-Sub, the LED stops flashing (it goes off), and Windows recognizes the wheel as new "USB input device" respectively as a "HID conform game controller", accompanied by its characteristical 'pling-pling' sounds.

So, have you followed this order?
1) Power on the wheel (with its own power supply)
2) Plug the teensy to USB (teensy is flashing quickly) 
3) Only after 1 and 2 connect the wheel to the 15 pin D-Sub (LED goes off and Windows recognizes the 'USB wheel')

For the software, have you tried the version 'adaptffbjoy-wheelFF_alpha1-Teensy.hex' in the download folder of the repo? It is the latest one that was released.

**Reply by Harbor-Gaming** [reply]
Yes, i did follow that order.
Btw, i went to the "downloads" folder in the repo and i fonly found these files:
adaptffbjoy-0.3.0(r54).hex
adaptffbjoy-0.5.0beta1-shift_4axis_Micro.hex 
adaptffbjoy-0.5.0beta1-shift_4axis_Teensy.hex
adaptffbjoy-0.5.0beta1.hex 
adaptffbjoy-breadboard.jpg
adaptffbjoy-circuit.png

i couldnt find the "adaptffbjoy-wheelFF_alpha1-Teensy.hex" do you have a direct link for it?
and anothre question, in the adaptffbjoy-circuit diagram, if u dont use the 4 trimpots, you connect the 4 "PF's" to GND or to VCC?   The EXACT circuit schematic shown works right? Cuz for some reason deepseek (i tried deepseek, chatgpt and grok ai to help me) told me the schematic is wrong, and told me to disconnect pf1 from anything and leave connected pf0 to +5v...
Btw, when i connect the adapter to the pc it starts flashing, when i then connect it to the wheel it continues flashing, and but then i can turn on or off the ffb button on the wheel, which if i turn on, it auto centers and starts doing force going to middle.

**Reply by You69Man** [reply]
Take a look into this repo folder (see also the link of @ej113 above under the heading "At last some firmware: ... I've put four firmware files here"):
https://github.com/ej113/adapt-ffb-joy/tree/micro_pinout_wheeltest2_LOG/downloads

There is also the circuit diagram that I used (without the optional trimpots and also without the optional capacitors, just the teensy and the three resistors):
https://github.com/ej113/adapt-ffb-joy/blob/micro_pinout_wheeltest2_LOG/downloads/adaptffbjoy-circuit.png

PF0, PF1, PF4, PF5 (trimpot options) are just left open on my circuitry.

**Reply by Harbor-Gaming** [reply]
Thank you! i found the hex file, i flashed the teensy 2.0 with it. 
But still the same situation.
The teensy keeps flashing and the pc doesnt detect it.  (And yes its connected to the wheel too and wheel has power).
Do you have ANY IDEA of what might be wrong or fixes to fix it?

**Reply by Harbor-Gaming** [reply]
yo, i noticed that when I connect the adapter first to the wheel and then to the pc, while its connected to the wheel it doesnt let me turn ffb on or off (it stays off all the time with no blinking) so the wheel is detecting it, cuz if i disconnect it the force button starts blinking again. and the teensy's led doesnt blink. but as soon as i connect it to the pc, the teensy starts blinking again and only then it lets me turn ffb on or off.
BTW, you used the 2x 2.2k ohm resistors and 1x 220ohm resistor and no capacitors (i used the 2 capacitors but instead of 1nF i used 1.5nF cuz that's what i had on hand, is it the cause of my problem?)
And also without the capacitors, did you also ommit the connection between before the resistors to GND?

**Reply by You69Man** [reply]
> yo, i noticed that when I connect the adapter first to the wheel and then to the pc, while its connected to the wheel ...

Yes, but in my opinion this is "undefined behaviour" because in this situation the adapter (teensy) is not really supplied accordingly but it is "backpowered" via the signal lines from the already powered on wheel.  
@ej113 provided an extensive description about this topic with a warning not to do so https://github.com/tloimu/adapt-ffb-joy/issues/49.

>BTW, you used the 2x 2.2k ohm resistors and 1x 220ohm resistor and no capacitors (i used the 2 capacitors but instead of 1nF i used 1.5nF cuz that's what i had on hand, is it the cause of my problem?) 

I don't know. Maybe its worth disconnecting them temporarily.

>And also without the capacitors, did you also ommit the connection between before the resistors to GND?

Which connection do you mean?
This is what should be connected (when omitting the "options"):

<img width="800" height="576" alt="Unbenannt" src="https://github.com/user-attachments/assets/404bb61a-57fb-4f51-bc83-d00548d47f4c" />

**Reply by You69Man** [reply]
If the circuitry is flawless I can only image an issue with the USB connection. I would try different USB cables, different USB ports (front, rear, USB2.0, USB3.0...), with / without USB hubs, maybe also on another PC. 
Maybe it's also a Windows USB driver issue. But as far as I know there is not special USB driver needed (or at least I don't remember to have installed any). 
Btw. I'm using Windows 10.

**Reply by Harbor-Gaming** [reply]
OMG, I REMOVED THOSE 2 LITTLE USELESS CAPACITORS CONNECTED TO THE GROUND AND THOSE BLOODY PLONKERS WHERE THE CAUSE OF THE PROBLEM!!! IT WORKS!!!! OMG THANK YOU SO MUCH FOR YOUR HELP!!! NOT EVEN CHATGPT, DEEPSEEK AND GROK AI ALL THE THREE OF THEM TOGETHER COULD HELP ME!!! IM SO GRATEFUL MAN THANK YOU!

**Reply by You69Man** [reply]
Congrats!! Glad I could help 😃!
